# Machine Learning Model Application

In this part, after merging the `GENERAL_STATS` dataset and `EMA_data` dataset, we will clean the data to create a dataset characterized by all the features and the target variable.

## Features Used for Training the Machine Learning Models:

- **B365H**: Betting odds for a home win (Bet365).
- **B365D**: Betting odds for a draw (Bet365).
- **B365A**: Betting odds for an away win (Bet365).
- **HTGD**: Goal difference for the home team up to the match.
- **ATGD**: Goal difference for the away team up to the match.
- **DiffPts**: Difference in total points between the two teams.
- **DiffFormPts**: Difference in recent form points between the two teams.
- **f_cornersAgainstHome**: Corners conceded by the home team (EMA feature).
- **f_cornersForHome**: Corners earned by the home team (EMA feature).
- **f_freesAgainstHome**: Fouls conceded by the home team (EMA feature).
- **f_freesForHome**: Fouls earned by the home team (EMA feature).
- **f_goalsAgainstHome**: Goals conceded by the home team (EMA feature).
- **f_goalsForHome**: Goals scored by the home team (EMA feature).
- **f_halfTimeGoalsAgainstHome**: Half-time goals conceded by the home team (EMA feature).
- **f_halfTimeGoalsForHome**: Half-time goals scored by the home team (EMA feature).
- **f_redsAgainstHome**: Red cards received by opponents of the home team (EMA feature).
- **f_redsForHome**: Red cards received by the home team (EMA feature).
- **f_shotsAgainstHome**: Shots taken by opponents of the home team (EMA feature).
- **f_shotsForHome**: Shots taken by the home team (EMA feature).
- **f_shotsOnTargetAgainstHome**: Shots on target by opponents of the home team (EMA feature).
- **f_shotsOnTargetForHome**: Shots on target by the home team (EMA feature).
- **f_yellowsAgainstHome**: Yellow cards received by opponents of the home team (EMA feature).
- **f_yellowsForHome**: Yellow cards received by the home team (EMA feature).
- **f_cornersAgainstAway**: Corners conceded by the away team (EMA feature).
- **f_cornersForAway**: Corners earned by the away team (EMA feature).
- **f_freesAgainstAway**: Fouls conceded by the away team (EMA feature).
- **f_freesForAway**: Fouls earned by the away team (EMA feature).
- **f_goalsAgainstAway**: Goals conceded by the away team (EMA feature).
- **f_goalsForAway**: Goals scored by the away team (EMA feature).
- **f_halfTimeGoalsAgainstAway**: Half-time goals conceded by the away team (EMA feature).
- **f_halfTimeGoalsForAway**: Half-time goals scored by the away team (EMA feature).
- **f_redsAgainstAway**: Red cards received by opponents of the away team (EMA feature).
- **f_redsForAway**: Red cards received by the away team (EMA feature).
- **f_shotsAgainstAway**: Shots taken by opponents of the away team (EMA feature).
- **f_shotsForAway**: Shots taken by the away team (EMA feature).
- **f_shotsOnTargetAgainstAway**: Shots on target by opponents of the away team (EMA feature).
- **f_shotsOnTargetForAway**: Shots on target by the away team (EMA feature).
- **f_yellowsAgainstAway**: Yellow cards received by opponents of the away team (EMA feature).
- **f_yellowsForAway**: Yellow cards received by the away team (EMA feature).

## Target Variable:

- **FTR**: 
  - `H`: Home Win  
  - `D`: Draw  
  - `A`: Away Win
lowsForAway'



In [160]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split, cross_val_score, StratifiedKFold, GridSearchCV, KFold
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.linear_model import LogisticRegression, LogisticRegressionCV
from sklearn.ensemble import RandomForestClassifier
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import roc_curve, roc_auc_score, classification_report, accuracy_score, confusion_matrix 
from xgboost import XGBClassifier
from sklearn.ensemble import StackingClassifier
from sklearn.model_selection import RandomizedSearchCV
from sklearn.pipeline import Pipeline
import warnings
import lightgbm as lgb

import os
import seaborn as sns
import matplotlib.pyplot as plt
from MachineLearningModels import*
warnings.filterwarnings('ignore')

%load_ext autoreload
%autoreload 2

DATA_PATH='data'

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


First of all we use discard the variables in the datasets we are sure we are not going to use in the analysis. Then we merge the datasets and filter them in order to use only the features to use in our machine learning models.

From the GENERAL_STATS dataset one we eliminate:
- 'FTHG', 'FTAG': goals in the match, we should predict the result so we don't know it
- 'HTGS', 'ATGS', 'HTGC','ATGC': goals done and conceded so far, already considered with the feature HTGD e ATGD
 - 'HTFormPts','ATFormPts': points home team and away tema in the last 5 mathces. Already considered DiffFormPts
 - 'HTP', 'ATP':points so far for home and away team so far. Already considered with DiffPts
 - 'MW', 'HTFormPtsStr', 'ATFormPtsStr'. Useless variables

Form the EMA_data dataset we discard: 'Unnamed: 0', 'f_DateHome', 'f_SeasonHome', 'HomeTeam', 'homeGame_x', 'f_DateAway', 'f_SeasonAway','AwayTeam', 'homeGame_y', 'gameId_y','gameId_x' which are either repetition of a variable in GENERAL_STATS or a useless variable

Then we merge both variable in order to have a unique dataset which contain all the features that will be used to predict the outcome of each match


In [162]:
general_stats = pd.read_csv('data/GENERAL_STATS.csv')

general_stats.columns

Index(['Date', 'Season', 'HomeTeam', 'AwayTeam', 'FTHG', 'FTAG', 'FTR', 'HTGS',
       'ATGS', 'HTGC', 'ATGC', 'HTP', 'ATP', 'B365H', 'B365D', 'B365A', 'MW',
       'gameId', 'HTFormPtsStr', 'ATFormPtsStr', 'HTFormPts', 'ATFormPts',
       'HTGD', 'ATGD', 'DiffPts', 'DiffFormPts'],
      dtype='object')

In [163]:
general_stats.drop(['FTHG', 'FTAG',
                 'HTGS','ATGS', 'HTGC', 'ATGC',  'HTFormPts', 
                 'ATFormPts', 'MW', 'HTFormPtsStr', 'ATFormPtsStr', 'HTP', 'ATP'], axis =1, inplace=True)
general_stats.columns

Index(['Date', 'Season', 'HomeTeam', 'AwayTeam', 'FTR', 'B365H', 'B365D',
       'B365A', 'gameId', 'HTGD', 'ATGD', 'DiffPts', 'DiffFormPts'],
      dtype='object')

In [164]:
ema_dataset = pd.read_csv("data/EMA_data.csv")

In [165]:
ema_dataset.drop(['Unnamed: 0', 'f_DateHome', 'f_SeasonHome', 'HomeTeam',
               'homeGame_x', 'f_DateAway', 'f_SeasonAway', 
               'AwayTeam', 'homeGame_y'], axis=1, inplace=True)


In [166]:
# Dataset merging w.r.t gameId
df = pd.merge(general_stats, ema_dataset, left_on='gameId', right_index=True) 
df.drop(['gameId_y','gameId_x'], axis=1, inplace=True)

df.head()

gameId        Date  Season  HomeTeam    AwayTeam FTR  B365H  B365D  B365A  \
0       1  2016-08-20    1617  Juventus  Fiorentina   H   1.50   4.20    7.0   
1       2  2016-08-20    1617      Roma     Udinese   H   1.44   4.50    7.5   
2       3  2016-08-21    1617  Atalanta       Lazio   A   3.10   3.25    2.4   
3       4  2016-08-21    1617   Bologna     Crotone   H   1.85   3.50    5.0   
4       5  2016-08-21    1617    Chievo       Inter   H   3.75   3.30    2.1   

   HTGD  ...  f_halfTimeGoalsAgainstAway  f_halfTimeGoalsForAway  \
0   1.0  ...                        1.00                    1.10   
1   4.0  ...                        0.55                    0.00   
2  -1.0  ...                        0.00                    1.35   
3   1.0  ...                        0.00                    0.55   
4   2.0  ...                        0.55                    1.55   

   f_redsAgainstAway  f_redsForAway  f_shotsAgainstAway  f_shotsForAway  \
0               0.45            0.0               10.45           20.00   
1               0.00            0.0               21.45           11.15   
2               0.00            0.0               15.70            8.25   
3               0.55            0.0               18.40           11.00   
4               1.00            0.0                7.45           11.15   

   f_shotsOnTargetAgainstAway  f_shotsOnTargetForAway  f_yellowsAgainstAway  \
0                        3.35                    7.85                  1.10   
1                        6.55                    4.10                  2.35   
2                        5.45                    3.25                  2.00   
3                        6.25                    2.10                  1.65   
4                        1.90                    4.45                  3.55   

   f_yellowsForAway  
0              1.90  
1              3.00  
2              1.45  
3              2.55  
4              2.35  

[5 rows x 45 columns]

In [167]:
df.columns

Index(['gameId', 'Date', 'Season', 'HomeTeam', 'AwayTeam', 'FTR', 'B365H',
       'B365D', 'B365A', 'HTGD', 'ATGD', 'DiffPts', 'DiffFormPts',
       'f_cornersAgainstHome', 'f_cornersForHome', 'f_freesAgainstHome',
       'f_freesForHome', 'f_goalsAgainstHome', 'f_goalsForHome',
       'f_halfTimeGoalsAgainstHome', 'f_halfTimeGoalsForHome',
       'f_redsAgainstHome', 'f_redsForHome', 'f_shotsAgainstHome',
       'f_shotsForHome', 'f_shotsOnTargetAgainstHome',
       'f_shotsOnTargetForHome', 'f_yellowsAgainstHome', 'f_yellowsForHome',
       'f_cornersAgainstAway', 'f_cornersForAway', 'f_freesAgainstAway',
       'f_freesForAway', 'f_goalsAgainstAway', 'f_goalsForAway',
       'f_halfTimeGoalsAgainstAway', 'f_halfTimeGoalsForAway',
       'f_redsAgainstAway', 'f_redsForAway', 'f_shotsAgainstAway',
       'f_shotsForAway', 'f_shotsOnTargetAgainstAway',
       'f_shotsOnTargetForAway', 'f_yellowsAgainstAway', 'f_yellowsForAway'],
      dtype='object')

In [168]:
df.to_csv(os.path.join(DATA_PATH, 'ML_data.csv'), index=False)

We save this dataset. We will use it later in order to analyze a feasible Betting strategy (Betting_Strategy.ipynb)

In [170]:
df.tail()

gameId        Date  Season    HomeTeam AwayTeam FTR  B365H  B365D  \
3061    3065  2024-08-31    2425       Lazio    Milan   D   2.90   3.40   
3062    3066  2024-08-31    2425      Napoli    Parma   H   1.48   4.33   
3063    3067  2024-09-01    2425  Fiorentina    Monza   D   1.67   3.90   
3064    3068  2024-09-01    2425       Genoa   Verona   A   2.00   3.30   
3065    3069  2024-09-01    2425    Juventus     Roma   D   1.75   3.60   

      B365A      HTGD  ...  f_halfTimeGoalsAgainstAway  \
3061    2.4  0.333333  ...                    0.835163   
3062    7.0  0.333333  ...                    0.606449   
3063    5.0  0.000000  ...                    0.807537   
3064    4.0 -0.333333  ...                    0.673632   
3065    5.0  2.000000  ...                    0.781684   

      f_halfTimeGoalsForAway  f_redsAgainstAway  f_redsForAway  \
3061                0.771618           0.187666       0.054997   
3062                1.095937           0.149224       0.000199   
3063                0.638397           0.167516       0.220112   
3064                1.281323           0.236926       0.016379   
3065                0.355927           0.003430       0.053792   

      f_shotsAgainstAway  f_shotsForAway  f_shotsOnTargetAgainstAway  \
3061           11.861504       13.153325                    4.682292   
3062           10.611429       16.771582                    3.787876   
3063           17.080906       10.258752                    5.552664   
3064           10.279474       17.192867                    4.329257   
3065           13.604644        6.440977                    4.027350   

      f_shotsOnTargetForAway  f_yellowsAgainstAway  f_yellowsForAway  
3061                4.268263              2.261853          2.022254  
3062                5.834891              1.663594          1.659606  
3063                4.050801              2.261122          1.986619  
3064                5.806002              1.383074          1.162247  
3065                2.425707              1.670801          2.401249  

[5 rows x 45 columns]

## Machine Learning Model Application

Now we prepare the dataset to correctly apply the machine learning models.

Since these are temporal variables, we cannot use the classic 70/30 or 80/20 random split of the data. Instead, we need to set a cutoff date to divide the data into training and test sets. We decided to use the following split:

- **Training data**: From the 16-17 season to the 22-23 season (inclusive - approximately 75% of the data).
- **Test data**: The 23-24 season and the 24-25 season (up to the last available match - approximately 25% of the data).

Next, all variables are normalized using the `StandardScaler`, and the categories of the target variable (FTR) are transformed as follows:

- **FTR**:  
  - `H`: Home Win == 2  
  - `D`: Draw == 1  
  - `A`: Away Win == 0

In [172]:
# Create training data: Include all seasons except 2324 and 2425
training_data = df.loc[~(df['Season'] == 2324) & ~(df['Season'] == 2425)].reset_index(drop=True)
# Create testing data: Include only seasons 2324 and 2425
testing_data = df.loc[(df['Season'] == 2324) | (df['Season'] == 2425)]

X = training_data.drop(['gameId', 'Date', 'Season', 'HomeTeam', 'AwayTeam',  'FTR'], axis=1)
Y = training_data['FTR']

X_test = testing_data.drop(['gameId', 'Date', 'Season', 'HomeTeam', 'AwayTeam', 'FTR'], axis=1)
y_test = testing_data['FTR']

print(f"Training features shape: {X.shape}")
print(f"Test features shape: {X_test.shape}")
print(f"Training labels shape: {Y.shape}")
print(f"Test labels shape: {y_test.shape}")

Training features shape: (2657, 39)
Test features shape: (409, 39)
Training labels shape: (2657,)
Test labels shape: (409,)


In [173]:
# Normalize the features and Trasform in numeric the Target variable
le = LabelEncoder()
scaler = StandardScaler()

Y = le.fit_transform(Y)      # Away Win = 0 , Draw = 1, # Home Win = 2
y_test = le.fit_transform(y_test)

X = scaler.fit_transform(X)
X_test = scaler.fit_transform(X_test)

# MachineLearningModels.py - Automated Pipeline for Model Training

The `MachineLearningModels.py` file implements an automated pipeline for training, validation, and optimization of machine learning models with the goal of reducing overfitting, maximizing performance, and comparing key metrics across models. The process also enables the saving of trained models in a dedicated directory (`ML_models`), making them reusable for future applications, such as in the next stage, `Betting_Strategy`.

The models considered include:

- **Logistic Regression (Simple)**: A baseline linear classification model.  
- **Elastic Net Logistic Regression**: An advanced variant of logistic regression that combines L1 regularization (sparsity) and L2 regularization (shrinkage), particularly useful when dealing with numerous correlated features, as in the given dataset.  
- **Random Forest**: A tree-based model ideal for exploring complex relationships between variables without requiring significant feature scaling.  
- **XGBoost (Extreme Gradient Boosting Classifier)**: A highly optimized and powerful gradient-boosting algorithm, designed for complex problems.  
- **LightGBM (Light Gradient Boosting Machine)**: A lightweight and faster alternative to XGBoost, optimized for handling large datasets.  

The primary goal is to identify the model with the best **F1-weighted score**, a balanced metric particularly suitable for datasets with imbalanced classes.

---

# Process Details

The `MachineLearningModels.py` script functions as follows:

1. **Model Initialization**  
   Each model is initialized with default or pre-defined parameters.

2. **Hyperparameter Optimization**  
   - **RandomizedSearchCV** is used to perform hyperparameter tuning by testing multiple random combinations of parameters.
   - For **Random Forest**, **XGBoost**, and **LightGBM**, this optimization is conducted in two phases:
     - **Broader Search**: An initial exploration of a wide range of parameter values.
     - **Deeper Search**: A more focused search around the best parameters from the broader search.

3. **Cross-Validation**  
   - Stratified 5-Fold Cross-Validation (`StratifiedKFold`) is employed to ensure robust evaluation of the models by maintaining the proportion of classes across folds.

4. **Metric Evaluation**  
   - Each model is evaluated using the **weighted F1-score**. The model with the highest weighted F1-score is selected.

5. **Model Saving**  
   - Trained models are saved as `.pkl` files in the `ML_models` directory to facilitate reuse without retraining.

6. **Comparison of Models**  
   - Models are compared based on their weighted F1-scores and other performance metrics (e.g., accuracy, precision, recall). The best model is selected and highlighted.

This systematic approach ensures the selection of an optimal model, balancing computational efficiency and predictive performance.



In [175]:
models = {
    "Logistic Regression": train_simple_logistic(X, Y, X_test, y_test),
    "Elastic Net Logistic": train_elastic_net(X, Y, X_test, y_test),
    "Random Forest": train_random_forest_improved(X, Y, X_test, y_test),
    "XGBoost": train_xgboost_improved(X, Y, X_test, y_test),
    "LightGBM": train_lightgbm_improved(X, Y, X_test, y_test),
}


Simple Logistic Regression:
Fitting 5 folds for each of 20 candidates, totalling 100 fits

Best parameters found:
{'C': 9.833308858067882, 'max_iter': 1000, 'solver': 'lbfgs'}

Best cross-validation scores:
accuracy: 0.615
f1_weighted: 0.581

Classification Report on test set:
              precision    recall  f1-score   support

           0       0.62      0.73      0.67       115
           1       0.35      0.11      0.17       124
           2       0.60      0.82      0.69       170

    accuracy                           0.58       409
   macro avg       0.52      0.56      0.51       409
weighted avg       0.53      0.58      0.53       409

Model saved to: ML_models\logistic_regression_simple.pkl
------------------------------------------------------------------------------------------

Elastic Net Logistic Regression:
Fitting 5 folds for each of 50 candidates, totalling 250 fits

Best parameters found:
{'C': 2.588799816000169, 'l1_ratio': 0.662522284353982, 'max_iter': 2000

In [176]:
best_model_name, best_model = evaluate_models(models, X_test, y_test)



Comparative Model Evaluation:

Logistic Regression:
              precision    recall  f1-score   support

           0       0.62      0.73      0.67       115
           1       0.35      0.11      0.17       124
           2       0.60      0.82      0.69       170

    accuracy                           0.58       409
   macro avg       0.52      0.56      0.51       409
weighted avg       0.53      0.58      0.53       409


Confusion Matrix:
[[ 84   9  22]
 [ 39  14  71]
 [ 13  17 140]]

Elastic Net Logistic:
              precision    recall  f1-score   support

           0       0.62      0.73      0.67       115
           1       0.37      0.12      0.18       124
           2       0.60      0.82      0.69       170

    accuracy                           0.58       409
   macro avg       0.53      0.56      0.52       409
weighted avg       0.54      0.58      0.53       409


Confusion Matrix:
[[ 84   9  22]
 [ 38  15  71]
 [ 13  17 140]]

Random Forest:
              pr

# Model Performance Analysis

## Logistic Regression
- **Precision**: High for `away win` (0.62) and `home win` (0.60), but low for `draw` (0.35).
- **Recall**: Very high for `home win` (0.82), but extremely low for `draw` (0.11), indicating the model struggles to correctly identify this outcome.
- **F1-Score**: Balanced for `away win` and `home win`, but very low for `draw` (0.17).
- **Conclusion**: Overall performance is acceptable, but the model struggles with imbalanced outcomes, particularly predicting `draw`.


## Elastic Net Logistic Regression
- Results are similar to the basic Logistic Regression:
  - Slight improvement in **Precision** and **Recall** for `draw` (`precision`: 0.35 → 0.37, `recall`: 0.11 → 0.12).
  - No significant improvement for the other outcomes.
- **Conclusion**: Adding L1 and L2 regularization provides minor gains but does not significantly outperform the base model.


## Random Forest
- **Precision**: Improved compared to Logistic Regression, especially for `draw` (0.42), and decent for the other two outcomes.
- **Recall**: Balanced overall (`away win`: 0.71, `draw`: 0.26, `home win`: 0.68).
- **F1-Score**: Higher across all outcomes compared to Logistic Regression. 
  - Best performance for `home win` (0.65).
- **Conclusion**: Random Forest performs better with imbalanced outcomes due to its ability to handle non-linear relationships and complex features.


## XGBoost
- **Precision** and **Recall**: Similar to Random Forest, with a slight drop in precision for `away win` and recall for `draw`.
- **F1-Score**: Comparable to Random Forest but slightly lower for `draw` (0.29 vs. 0.32).
- **Conclusion**: Although powerful, XGBoost does not outperform Random Forest on this dataset, likely due to limited hyperparameter tuning.


## LightGBM
- **Precision**: Similar to XGBoost, but slightly lower across all outcomes.
- **Recall**: Good for `home win` (0.72), but low for `draw` (0.22).
- **F1-Score**: Comparable to XGBoost and Random Forest but worse for `draw`.
- **Conclusion**: While LightGBM offers excellent training speed, its overall performance is weaker compared to Random Forest.


## Confusion Matrix Insights
1. **Away Win**: Identified with high precision and recall across all models.
2. **Draw**: All models struggle to identify this outcome correctly, with high false negative rates.
3. **Home Win**: Consistently well-classified, with high recall across all models.


## General Observations
1. **Weighted F1-Score**: Random Forest (0.544) emerges as the best model in terms of balancing precision and recall, delivering the highest overall performance.
2. **Class Imbalance**: The `draw` outcome poses a significant challenge for all models.
3. **Model Selection**: Random Forest offers the best balance between simplicity and performance. While XGBoost and LightGBM may benefit from more aggressive hyperparameter tuning, it's uncertain whether the additional computational resources and time required would significantly enhance their predictive power.




--> Proceed to the "Betting_Strategy" section.